File: mobile/iosApp/Bissbilanz/Models/Supplement.swift
import Foundation

enum ScheduleType: String, Codable {
    case daily
    case everyOtherDay = "every_other_day"
    case weekly
    case specificDays = "specific_days"
}

enum FoodKind: String, Codable {
    case food
    case supplement
}

struct SupplementBackingFood: Codable, Identifiable {
    let id: String
    let name: String
    let brand: String?
    let kind: FoodKind
    let servingSize: Double
    let servingUnit: String
    let calories: Double
    let protein: Double
    let carbs: Double
    let fat: Double
    let fiber: Double
    let ingredientsText: String?
}

struct Supplement: Codable, Identifiable {
    let id: String
    let userId: String
    let name: String
    let scheduleType: ScheduleType
    let scheduleDays: [Int]?
    let scheduleStartDate: String?
    let isActive: Bool
    let sortOrder: Int
    let timeOfDay: String?
    let createdAt: String?
    let updatedAt: String?
    let ingredients: [SupplementIngredient]
}

struct SupplementIngredient: Codable, Identifiable {
    let id: String
    let supplementId: String
    let foodId: String
    let servings: Double
    let sortOrder: Int
    let food: SupplementBackingFood
}

struct SupplementLog: Codable {
    let supplementId: String
    let date: String
    let takenAt: String
    let entryIds: [String]
}

struct SupplementChecklist: Codable, Identifiable {
    let supplement: Supplement
    let taken: Bool
    let takenAt: String?

    var id: String { supplement.id }
}

struct SupplementsResponse: Codable {
    let supplements: [Supplement]
}

struct SupplementResponse: Codable {
    let supplement: Supplement
}

struct SupplementLogResponse: Codable {
    let log: SupplementLog
}

struct SupplementChecklistResponse: Codable {
    let checklist: [SupplementChecklist]
}

struct SupplementCreate: Codable {
    let name: String
    let scheduleType: ScheduleType
    var scheduleDays: [Int]?
    var scheduleStartDate: String?
    var isActive: Bool?
    var sortOrder: Int?
    var timeOfDay: String?
    let ingredients: [SupplementIngredientInput]
}

struct SupplementUpdate: Codable {
    var name: String?
    var scheduleType: ScheduleType?
    var scheduleDays: [Int]?
    var scheduleStartDate: String?
    var isActive: Bool?
    var sortOrder: Int?
    var timeOfDay: String?
    var ingredients: [SupplementIngredientInput]?
}

struct SupplementIngredientInput: Codable {
    var foodId: String?
    var food: SupplementBackingFoodInput?
    var servings: Double?
    var sortOrder: Int?
}

// Inline backing food payload — kept minimal to match the web form; the server
// normalises any missing macro fields to zero.
struct SupplementBackingFoodInput: Codable {
    let name: String
    let servingSize: Double
    let servingUnit: String
    let calories: Double
    let protein: Double
    let carbs: Double
    let fat: Double
    let fiber: Double
    var ingredientsText: String?
}

File: mobile/shared/src/commonMain/kotlin/com/bissbilanz/model/Supplement.kt
package com.bissbilanz.model

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

@Serializable
enum class ScheduleType {
    @SerialName("daily")
    DAILY,

    @SerialName("every_other_day")
    EVERY_OTHER_DAY,

    @SerialName("weekly")
    WEEKLY,

    @SerialName("specific_days")
    SPECIFIC_DAYS,
}

@Serializable
data class Supplement(
    val id: String,
    val userId: String,
    val name: String,
    val dosage: Double,
    val dosageUnit: String,
    val scheduleType: ScheduleType,
    val scheduleDays: List<Int>? = null,
    val scheduleStartDate: String? = null,
    val isActive: Boolean = true,
    val sortOrder: Int = 0,
    val timeOfDay: String? = null,
    val createdAt: String? = null,
    val updatedAt: String? = null,
    val ingredients: List<SupplementIngredient>? = null,
)

@Serializable
data class SupplementIngredient(
    val id: String? = null,
    val supplementId: String? = null,
    val name: String,
    val dosage: Double,
    val dosageUnit: String,
    val sortOrder: Int = 0,
)

@Serializable
data class SupplementCreate(
    val name: String,
    val dosage: Double,
    val dosageUnit: String,
    val scheduleType: ScheduleType,
    val scheduleDays: List<Int>? = null,
    val scheduleStartDate: String? = null,
    val isActive: Boolean? = null,
    val sortOrder: Int? = null,
    val timeOfDay: String? = null,
    val ingredients: List<SupplementIngredientInput>? = null,
)

@Serializable
data class SupplementIngredientInput(
    val name: String,
    val dosage: Double,
    val dosageUnit: String,
    val sortOrder: Int? = null,
)

@Serializable
data class SupplementLog(
    val id: String,
    val supplementId: String,
    val userId: String,
    val date: String,
    val takenAt: String,
    val createdAt: String? = null,
)

@Serializable
data class SupplementHistoryEntry(
    val log: SupplementHistoryLog,
    val supplementName: String,
    val dosage: Double,
    val dosageUnit: String,
)

@Serializable
data class SupplementHistoryLog(
    val id: String,
    val supplementId: String,
    val userId: String,
    val date: String,
    val takenAt: String,
)

@Serializable
data class SupplementHistoryResponse(
    val history: List<SupplementHistoryEntry>,
)

@Serializable
data class SupplementsResponse(
    val supplements: List<Supplement>,
)
